version-ranges/src/lib.rs

```diff
@@ -826,6 +826,87 @@ impl<V: Ord + Clone> Ranges<V> {
+        let mut segments: SmallVec<[Interval<V>; 1]> = SmallVec::new();
+
+        for segment in iter {
+            if !valid_segment(&segment.start_bound(), &segment.end_bound()) {
+                continue;
+            }
+            // Find where to insert the new segment
+            let insertion_point = segments.partition_point(|elem: &Interval<V>| {
+                cmp_bounds_start(elem.start_bound(), segment.start_bound())
+                    .unwrap()
+                    .is_lt()
+            });
+            // Is it overlapping with the previous segment?
+            let previous_overlapping = insertion_point > 0
+                && !end_before_start_with_gap(
+                    &segments[insertion_point - 1].end_bound(),
+                    &segment.start_bound(),
+                );
+
+            // Is it overlapping with the following segment?
+            let next_overlapping = insertion_point < segments.len()
+                && !end_before_start_with_gap(
+                    &segment.end_bound(),
+                    &segments[insertion_point].start_bound(),
+                );
+
+            match (previous_overlapping, next_overlapping) {
+                (true, true) => {
+                    // previous:  |------|
+                    // segment:       |------|
+                    // following:          |------|
+                    //
+                    // final:     |---------------|
+                    // We merge all three segments into one, which is effectively removing one of
+                    // two previously inserted and changing the bounds on the other.
+                    let following = segments.remove(insertion_point);
+                    segments[insertion_point - 1].1 = following.1;
+                }
+                (true, false) => {
+                    // previous:  |------|
+                    // segment:       |------|
+                    // following:                |------|
+                    //
+                    // final:     |----------|   |------|
+                    // We can reuse the existing element by extending it.
+                    segments[insertion_point - 1].1 = segment.1;
+                }
+                (false, true) => {
+                    // previous:  |------|
+                    // segment:             |------|
+                    // following:               |------|
+                    //
+                    // final:    |------|   |---------|
+                    // We can reuse the existing element by extending it.
+                    segments[insertion_point].0 = segment.0;
+                }
+                (false, false) => {
+                    // previous:  |------|
+                    // segment:             |------|
+                    // following:                      |------|
+                    //
+                    // final:    |------|   |------|   |------|
+                    segments.insert(insertion_point, segment);
```

Comment:
This is doing insertion sort. Witch is `O(n^2)`, because this `insert` (or the `remove` above) is `O(n)`. Shoving all the segments into a heap should get us `O(n*log(n))` time, but take `O(n)` space. I don't know if it actually matters in practice.

If the point is to be practical I'm happy to merge this with a comment about alternative implementations. If the point of this is to implement the algorithmically best solution, we can do better than this.

Reply:
One solution would be collecting the segments into a sorted datastructure, then building a second ranges without any shifting around: We pay the price of two datastructure for no moving around. This solution works with only a single ranges, inserting or removing elements as we detect overlaps: If the elements are already in order, it's only pushes to a single smallvec (I chose not to reserve, since we could vastly over-reserve). As usual, I'd go with this implementation and change it if benchmarking with users shows that another algorithm would be better.

Reply:
Can we add a comment  on these two lines explaining that the call is `O(n)`  which makes the algorithm `O(n^2)`  but it should be good enough for now. That way  somebody looking at a flame graph that's hot here  has some clue that there is  a fix we've already thought of.